File: Sources/RedlineCore/Snapshot.swift
// Wire format between the menu bar app and the widget. A widget cannot poll or parse
// transcripts inside its time budget, so the app writes this and the widget only renders it.
import Foundation

public struct Snapshot: Codable, Equatable {
    public struct Window: Codable, Equatable, Identifiable {
        // Providers share window keys, so identity must include the provider or SwiftUI
        // renders one row twice and drops the other.
        public var id: String { "\(provider)|\(key)" }

        public let provider: String
        public let key: String
        public let utilization: Double
        public let resetsAt: Date?

        public var displayName: String {
            switch key {
            case "five_hour":        return "Session · 5h"
            case "seven_day":        return provider == "Claude" ? "Week · all models" : "Week"
            case "seven_day_opus":   return "Week · Opus"
            case "seven_day_sonnet": return "Week · Sonnet"
            default: return key.replacingOccurrences(of: "_", with: " ")
            }
        }

        public init(provider: String, key: String, utilization: Double, resetsAt: Date?) {
            self.provider = provider
            self.key = key
            self.utilization = utilization
            self.resetsAt = resetsAt
        }
    }

    public struct Totals: Codable, Equatable {
        public let io: Int
        public let cost: Double
        public let hasUnpriced: Bool

        public init(io: Int, cost: Double, hasUnpriced: Bool) {
            self.io = io
            self.cost = cost
            self.hasUnpriced = hasUnpriced
        }

        public init(_ agg: Agg) {
            self.init(io: agg.io, cost: agg.cost, hasUnpriced: agg.hasUnpriced)
        }
    }

    /// Live Ollama state, carried in the snapshot so the sandboxed widget never needs network
    /// access of its own.
    public struct Ollama: Codable, Equatable {
        public struct Running: Codable, Equatable {
            public let name: String
            public let sizeBytes: Int64
            public let vramShare: Double

            public init(name: String, sizeBytes: Int64, vramShare: Double) {
                self.name = name
                self.sizeBytes = sizeBytes
                self.vramShare = vramShare
            }
        }

        public let reachable: Bool
        public let version: String?
        public let running: [Running]
        public let downloadedCount: Int
        public let downloadedBytes: Int64

        public init(reachable: Bool, version: String?, running: [Running],
                    downloadedCount: Int, downloadedBytes: Int64) {
            self.reachable = reachable
            self.version = version
            self.running = running
            self.downloadedCount = downloadedCount
            self.downloadedBytes = downloadedBytes
        }
    }

    public let updatedAt: Date
    public let limits: [Window]
    public let today: Totals
    public let week: Totals
    // Optional so an older snapshot on disk still decodes rather than leaving the widget blank
    public let todayByProvider: [String: Totals]?
    public let weekByProvider: [String: Totals]?
    public let ollama: Ollama?

    // Timestamps are stored at whole-second precision, since the JSON encoding is ISO8601
    // and nothing here is finer-grained than a poll interval.
    /// A provider's own reported health, from its public status feed. Optional end to end
    /// so snapshots written before this field existed still decode.
    public struct Service: Codable, Equatable {
        public let provider: String
        public let indicator: String
        public let description: String

        public init(provider: String, indicator: String, description: String) {
            self.provider = provider
            self.indicator = indicator
            self.description = description
        }

        public var isOperational: Bool { indicator == "none" || indicator == "local" }

        /// Calm and factual: report what the operator reports, never alarm. The two local
        /// indicators are Ollama's, probed directly rather than read from a status page.
        public var phrase: String {
            switch indicator {
            case "none":  return "service ok"
            case "minor": return "minor incident reported"
            case "major", "critical": return "outage reported"
            case "local": return "local server running"
            case "local-down": return "local server not reachable"
            default: return "status unknown"
            }
        }
    }

    public var services: [Service]?

    public init(updatedAt: Date, limits: [LimitWindow], today: Agg, week: Agg,
                ollama: Ollama? = nil, services: [Service]? = nil) {
        self.updatedAt = Snapshot.truncate(updatedAt)
        self.todayByProvider = today.providers.mapValues {
            Totals(io: $0.io, cost: $0.cost, hasUnpriced: today.hasUnpriced)
        }
        self.weekByProvider = week.providers.mapValues {
            Totals(io: $0.io, cost: $0.cost, hasUnpriced: week.hasUnpriced)
        }
        self.ollama = ollama
        self.limits = limits.map {
            Window(provider: $0.provider, key: $0.key, utilization: $0.utilization,
                   resetsAt: $0.resetsAt.map(Snapshot.truncate))
        }
        self.today = Totals(today)
        self.week = Totals(week)
        self.services = services
    }

    static func truncate(_ d: Date) -> Date {
        Date(timeIntervalSince1970: d.timeIntervalSince1970.rounded(.down))
    }

    // Widgets refresh on the system's schedule, not ours, so a reading can be minutes old.
    // Callers must surface staleness rather than implying the number is live.
    public func isStale(now: Date = Date(), tolerance: TimeInterval = 900) -> Bool {
        now.timeIntervalSince(updatedAt) > tolerance
    }

    public func worst(prefix: String, provider: String? = nil) -> Window? {
        limits
            .filter { $0.key.hasPrefix(prefix) }
            .filter { provider == nil
                      || $0.provider.caseInsensitiveCompare(provider!) == .orderedSame }
            .max(by: { $0.utilization < $1.utilization })
    }

    public func windows(for provider: String) -> [Window] {
        limits.filter { $0.provider.caseInsensitiveCompare(provider) == .orderedSame }
    }

    public func today(for provider: String?) -> Totals {
        guard let provider else { return today }
        return todayByProvider?[provider] ?? Totals(io: 0, cost: 0, hasUnpriced: false)
    }

    public func week(for provider: String?) -> Totals {
        guard let provider else { return week }
        return weekByProvider?[provider] ?? Totals(io: 0, cost: 0, hasUnpriced: false)
    }
}

// Reads and writes the snapshot in a container both processes can reach. Falls back to a
// per-user path when no App Group is configured, so the app works before the widget exists.
public enum SnapshotStore {
    public static let appGroup = "group.com.goriparthi.redline"
    public static let fileName = "snapshot.json"

    public static let widgetBundleID = "com.goriparthi.redline.widget"

    /// Inside a sandboxed extension this resolves to its own container, which it can always
    /// read whatever the signing situation. In the app it resolves to the normal location.
    public static var localAppSupportURL: URL? {
        FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)
            .first?
            .appendingPathComponent("redline/\(fileName)")
    }

    /// The widget's container, addressed absolutely so the app can write into it. This is the
    /// only route that works for an ad-hoc signed sandboxed widget: App Group containers need
    /// a real Team ID, and a path exception outside the container is not granted.
    public static var widgetContainerURL: URL {
        FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent(
                "Library/Containers/\(widgetBundleID)/Data/Library/Application Support/redline/\(fileName)")
    }

    // Always available, and readable by any non-sandboxed process of this user
    public static var userURL: URL {
        FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent(".local/share/redline/\(fileName)")
    }

    // Only resolves for code signed with a real Team ID. An ad-hoc signed sandboxed
    // extension gets nil here, which is why the user path exists as the primary.
    public static func groupURL(appGroup group: String? = appGroup) -> URL? {
        guard let group,
              let container = FileManager.default
                .containerURL(forSecurityApplicationGroupIdentifier: group) else { return nil }
        return container.appendingPathComponent(fileName)
    }

    public static func url(appGroup group: String? = appGroup) -> URL {
        groupURL(appGroup: group) ?? userURL
    }

    // Every location the snapshot is written to. The widget container is included only when
    // it already exists, so nothing is created for a widget that was never installed.
    public static var writeTargets: [URL] {
        var out = [userURL]
        let container = widgetContainerURL
        if FileManager.default.fileExists(
            atPath: container.deletingLastPathComponent()
                .deletingLastPathComponent().path) {
            out.append(container)
        }
        if let g = groupURL() { out.append(g) }
        return out
    }

    // Own container first: in the widget that is the one guaranteed readable location
    public static var readCandidates: [URL] {
        var out: [URL] = []
        if let local = localAppSupportURL { out.append(local) }
        if let g = groupURL() { out.append(g) }
        out.append(userURL)
        return out
    }

    @discardableResult
    public static func writeEverywhere(_ snapshot: Snapshot) -> Bool {
        // Succeed if any target takes it; a missing App Group must not fail the write
        writeTargets.map { write(snapshot, to: $0) }.contains(true)
    }

    public static func readAny() -> Snapshot? {
        for url in readCandidates {
            if let s = read(from: url) { return s }
        }
        return nil
    }

    private static var encoder: JSONEncoder {
        let e = JSONEncoder()
        e.dateEncodingStrategy = .iso8601
        return e
    }

    private static var decoder: JSONDecoder {
        let d = JSONDecoder()
        d.dateDecodingStrategy = .iso8601
        return d
    }

    @discardableResult
    public static func write(_ snapshot: Snapshot, to url: URL = url()) -> Bool {
        guard let data = try? encoder.encode(snapshot) else { return false }
        do {
            try FileManager.default.createDirectory(
                at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
            try data.write(to: url, options: .atomic)
            // Usage and cost figures are nobody else's business on a shared machine
            try? FileManager.default.setAttributes([.posixPermissions: 0o600],
                                                   ofItemAtPath: url.path)
            return true
        } catch {
            return false
        }
    }

    public static func read(from url: URL = url()) -> Snapshot? {
        guard let data = try? Data(contentsOf: url) else { return nil }
        return try? decoder.decode(Snapshot.self, from: data)
    }
}

File: windows/src/Redline.Core/Snapshot.cs
// Wire format between the app and the widget. The widget cannot parse transcripts in its time
// budget, so the app writes this and the widget only renders it. Keys match Swift's Codable output.
using System.Globalization;
using System.Text.Json.Nodes;

namespace Redline.Core;

public sealed class Snapshot : IEquatable<Snapshot>
{
    public sealed record Window(string Provider, string Key, double Utilization, DateTimeOffset? ResetsAt)
    {
        // Providers share window keys, so identity must include the provider
        public string Id => $"{Provider}|{Key}";

        public string DisplayName => Key switch
        {
            "five_hour" => "Session · 5h",
            "seven_day" => Provider == "Claude" ? "Week · all models" : "Week",
            "seven_day_opus" => "Week · Opus",
            "seven_day_sonnet" => "Week · Sonnet",
            _ => Key.Replace('_', ' '),
        };
    }

    public sealed record Totals(int Io, double Cost, bool HasUnpriced)
    {
        public static Totals From(Agg agg) => new(agg.Io, agg.Cost, agg.HasUnpriced);
        internal static readonly Totals Zero = new(0, 0, false);
    }

    /// Live Ollama state, carried here so the widget never needs network access of its own.
    /// Swift's Snapshot.Ollama; renamed because C# cannot share a name with the property.
    public sealed record OllamaSection(bool Reachable, string? Version, IReadOnlyList<OllamaSection.RunningModel> Running,
                                       int DownloadedCount, long DownloadedBytes, int? CloudCount = null)
    {
        public sealed record RunningModel(string Name, long SizeBytes, double VramShare);

        /// Models actually stored on this disk, with cloud entries taken back out
        public int LocalCount => DownloadedCount - (CloudCount ?? 0);

        public bool Equals(OllamaSection? o) => o is not null && Reachable == o.Reachable &&
            Version == o.Version && Running.SequenceEqual(o.Running) && DownloadedCount == o.DownloadedCount &&
            DownloadedBytes == o.DownloadedBytes && CloudCount == o.CloudCount;
        public override int GetHashCode() => HashCode.Combine(Reachable, Version, DownloadedCount, DownloadedBytes);
    }

    /// A provider's own reported health, from its public status feed.
    public sealed record Service(string Provider, string Indicator, string Description)
    {
        public bool IsOperational => Indicator is "none" or "local";

        /// Calm and factual. The two local indicators are Ollama's, probed directly.
        public string Phrase => Indicator switch
        {
            "none" => "service ok",
            "minor" => "minor incident reported",
            "major" or "critical" => "outage reported",
            "local" => "local server running",
            "local-down" => "local server not reachable",
            _ => "status unknown",
        };
    }

    public DateTimeOffset UpdatedAt { get; }
    public IReadOnlyList<Window> Limits { get; }
    /// When Claude's windows were last true; the feed only writes while Claude Code runs.
    public DateTimeOffset? ClaudeLimitsAsOf { get; }
    public Totals Today { get; }
    public Totals Week { get; }
    // Optional so an older snapshot on disk still decodes rather than leaving the widget blank
    public IReadOnlyDictionary<string, Totals>? TodayByProvider { get; }
    public IReadOnlyDictionary<string, Totals>? WeekByProvider { get; }
    public OllamaSection? Ollama { get; }
    public List<Service>? Services { get; set; }

    public Snapshot(DateTimeOffset updatedAt, IEnumerable<LimitWindow> limits, Agg today, Agg week,
                    OllamaSection? ollama = null, List<Service>? services = null,
                    DateTimeOffset? claudeLimitsAsOf = null)
    {
        UpdatedAt = Truncate(updatedAt);
        ClaudeLimitsAsOf = claudeLimitsAsOf is { } c ? Truncate(c) : null;
        TodayByProvider = today.Providers.ToDictionary(kv => kv.Key, kv => new Totals(kv.Value.Io, kv.Value.Cost, today.HasUnpriced));
        WeekByProvider = week.Providers.ToDictionary(kv => kv.Key, kv => new Totals(kv.Value.Io, kv.Value.Cost, week.HasUnpriced));
        Ollama = ollama;
        Limits = limits.Select(w => new Window(w.Provider, w.Key, w.Utilization,
                                               w.ResetsAt is { } r ? Truncate(r) : null)).ToList();
        Today = Totals.From(today);
        Week = Totals.From(week);
        Services = services;
    }

    private Snapshot(DateTimeOffset updatedAt, List<Window> limits, DateTimeOffset? claudeLimitsAsOf,
                     Totals today, Totals week, Dictionary<string, Totals>? todayByProvider,
                     Dictionary<string, Totals>? weekByProvider, OllamaSection? ollama, List<Service>? services)
    {
        UpdatedAt = updatedAt;
        Limits = limits;
        ClaudeLimitsAsOf = claudeLimitsAsOf;
        Today = today;
        Week = week;
        TodayByProvider = todayByProvider;
        WeekByProvider = weekByProvider;
        Ollama = ollama;
        Services = services;
    }

    // Whole-second precision, since the ISO8601 encoding carries nothing finer
    internal static DateTimeOffset Truncate(DateTimeOffset d) =>
        DateTimeOffset.FromUnixTimeSeconds((long)Math.Floor(d.ToUnixTimeMilliseconds() / 1000.0));

    /// Widgets refresh on the system's schedule, so callers must surface staleness.
    public bool IsStale(DateTimeOffset? now = null, double tolerance = 900) =>
        ((now ?? DateTimeOffset.UtcNow) - UpdatedAt).TotalSeconds > tolerance;

    /// Claude's windows age apart from the snapshot; stale ones render drained of status colour.
    public bool ClaudeLimitsAreStale(DateTimeOffset? now = null, double tolerance = 900)
    {
        if (ClaudeLimitsAsOf is not { } at) return false;
        return ((now ?? DateTimeOffset.UtcNow) - at).TotalSeconds > tolerance;
    }

    public Window? Worst(string prefix, string? provider = null) =>
        Limits.Where(w => w.Key.StartsWith(prefix, StringComparison.Ordinal))
              .Where(w => provider is null || string.Equals(w.Provider, provider, StringComparison.OrdinalIgnoreCase))
              .Aggregate((Window?)null, (best, w) => best is null || w.Utilization > best.Utilization ? w : best);

    public List<Window> Windows(string provider) =>
        Limits.Where(w => string.Equals(w.Provider, provider, StringComparison.OrdinalIgnoreCase)).ToList();

    /// Swift's today(for:); a C# method cannot share the Today property's name. Null means all.
    public Totals TodayFor(string? provider) =>
        provider is null ? Today : TodayByProvider?.GetValueOrDefault(provider) ?? Totals.Zero;

    public Totals WeekFor(string? provider) =>
        provider is null ? Week : WeekByProvider?.GetValueOrDefault(provider) ?? Totals.Zero;

    public bool Equals(Snapshot? o) => o is not null && UpdatedAt == o.UpdatedAt &&
        Limits.SequenceEqual(o.Limits) && ClaudeLimitsAsOf == o.ClaudeLimitsAsOf &&
        Today == o.Today && Week == o.Week && DictEquals(TodayByProvider, o.TodayByProvider) &&
        DictEquals(WeekByProvider, o.WeekByProvider) && Equals(Ollama, o.Ollama) &&
        (Services is null ? o.Services is null : o.Services is not null && Services.SequenceEqual(o.Services));

    public override bool Equals(object? obj) => Equals(obj as Snapshot);
    public override int GetHashCode() => HashCode.Combine(UpdatedAt, Limits.Count, Today, Week);

    private static bool DictEquals(IReadOnlyDictionary<string, Totals>? a, IReadOnlyDictionary<string, Totals>? b)
    {
        if (a is null || b is null) return a is null && b is null;
        return a.Count == b.Count && a.All(kv => b.TryGetValue(kv.Key, out var v) && v == kv.Value);
    }

    // JSON, in the exact shape Swift's JSONEncoder writes: camelCase keys, nil omitted, ISO8601 dates

    internal static string Iso(DateTimeOffset d) => DiagnosticsLog.Stamp(d);

    internal static DateTimeOffset? ParseIso(string? s)
    {
        if (s is null) return null;
        return DateTimeOffset.TryParseExact(s, "yyyy-MM-dd'T'HH:mm:ssK", CultureInfo.InvariantCulture,
            DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal, out var d) ? d : null;
    }

    public JsonObject ToJson()
    {
        var o = new JsonObject
        {
            ["updatedAt"] = Iso(UpdatedAt),
            ["limits"] = new JsonArray(Limits.Select(w =>
            {
                var j = new JsonObject { ["provider"] = w.Provider, ["key"] = w.Key, ["utilization"] = w.Utilization };
                if (w.ResetsAt is { } r) j["resetsAt"] = Iso(r);
                return (JsonNode)j;
            }).ToArray()),
        };
        if (ClaudeLimitsAsOf is { } asOf) o["claudeLimitsAsOf"] = Iso(asOf);
        o["today"] = TotalsJson(Today);
        o["week"] = TotalsJson(Week);
        if (TodayByProvider is not null) o["todayByProvider"] = MapJson(TodayByProvider);
        if (WeekByProvider is not null) o["weekByProvider"] = MapJson(WeekByProvider);
        if (Ollama is { } ol)
        {
            var j = new JsonObject { ["reachable"] = ol.Reachable };
            if (ol.Version is not null) j["version"] = ol.Version;
            j["running"] = new JsonArray(ol.Running.Select(r => (JsonNode)new JsonObject
            {
                ["name"] = r.Name, ["sizeBytes"] = r.SizeBytes, ["vramShare"] = r.VramShare,
            }).ToArray());
            j["downloadedCount"] = ol.DownloadedCount;
            j["downloadedBytes"] = ol.DownloadedBytes;
            if (ol.CloudCount is { } cc) j["cloudCount"] = cc;
            o["ollama"] = j;
        }
        if (Services is not null)
            o["services"] = new JsonArray(Services.Select(s => (JsonNode)new JsonObject
            {
                ["provider"] = s.Provider, ["indicator"] = s.Indicator, ["description"] = s.Description,
            }).ToArray());
        return o;
    }

    private static JsonObject TotalsJson(Totals t) =>
        new() { ["io"] = t.Io, ["cost"] = t.Cost, ["hasUnpriced"] = t.HasUnpriced };

    private static JsonObject MapJson(IReadOnlyDictionary<string, Totals> map)
    {
        var o = new JsonObject();
        foreach (var (k, v) in map.OrderBy(kv => kv.Key, StringComparer.Ordinal)) o[k] = TotalsJson(v);
        return o;
    }

    private sealed class DecodeError : Exception
    {
        public DecodeError(string key) : base($"missing or mistyped key: {key}") { }
    }

    private static T Req<T>(T? v, string key) where T : struct => v ?? throw new DecodeError(key);
    private static T Req<T>(T? v, string key, int _ = 0) where T : class => v ?? throw new DecodeError(key);

    /// Throws when a key Swift's decoder requires is missing or mistyped; optional keys may be absent.
    public static Snapshot FromJson(JsonObject o)
    {
        var limits = Req(o["limits"] as JsonArray, "limits").Select(n =>
        {
            var w = Req(n as JsonObject, "limits[]");
            return new Window(Req(Json.Str(w["provider"]), "provider"), Req(Json.Str(w["key"]), "key"),
                              Req(Json.Num(w["utilization"]), "utilization"), OptDate(w, "resetsAt"));
        }).ToList();
        OllamaSection? ollama = null;
        if (Present(o, "ollama"))
        {
            var j = Req(o["ollama"] as JsonObject, "ollama");
            var running = Req(j["running"] as JsonArray, "running").Select(n =>
            {
                var r = Req(n as JsonObject, "running[]");
                return new OllamaSection.RunningModel(Req(Json.Str(r["name"]), "name"),
                    Req(Json.Long(r["sizeBytes"]), "sizeBytes"), Req(Json.Num(r["vramShare"]), "vramShare"));
            }).ToList();
            ollama = new OllamaSection(Req(Json.Bool(j["reachable"]), "reachable"),
                Present(j, "version") ? Req(Json.Str(j["version"]), "version") : null, running,
                Req(Json.Int(j["downloadedCount"]), "downloadedCount"),
                Req(Json.Long(j["downloadedBytes"]), "downloadedBytes"),
                Present(j, "cloudCount") ? Req(Json.Int(j["cloudCount"]), "cloudCount") : null);
        }
        List<Service>? services = null;
        if (Present(o, "services"))
            services = Req(o["services"] as JsonArray, "services").Select(n =>
            {
                var s = Req(n as JsonObject, "services[]");
                return new Service(Req(Json.Str(s["provider"]), "provider"),
                    Req(Json.Str(s["indicator"]), "indicator"), Req(Json.Str(s["description"]), "description"));
            }).ToList();
        return new Snapshot(Req(ParseIso(Json.Str(o["updatedAt"])), "updatedAt"), limits,
            OptDate(o, "claudeLimitsAsOf"), TotalsFrom(o["today"], "today"), TotalsFrom(o["week"], "week"),
            MapFrom(o, "todayByProvider"), MapFrom(o, "weekByProvider"), ollama, services);
    }

    // Swift's decodeIfPresent treats an explicit null like an absent key
    private static bool Present(JsonObject o, string key) => o[key] is not null;

    private static DateTimeOffset? OptDate(JsonObject o, string key) =>
        Present(o, key) ? Req(ParseIso(Json.Str(o[key])), key) : null;

    private static Totals TotalsFrom(JsonNode? n, string key)
    {
        var t = Req(n as JsonObject, key);
        return new Totals(Req(Json.Int(t["io"]), "io"), Req(Json.Num(t["cost"]), "cost"),
                          Req(Json.Bool(t["hasUnpriced"]), "hasUnpriced"));
    }

    private static Dictionary<string, Totals>? MapFrom(JsonObject o, string key)
    {
        if (!Present(o, key)) return null;
        return Req(o[key] as JsonObject, key).ToDictionary(kv => kv.Key, kv => TotalsFrom(kv.Value, key));
    }
}

/// Reads and writes the snapshot where the app and the widget can both reach it. Windows has no
/// App Group or widget sandbox, so the user path under ~/.local/share/redline carries it.
public static class SnapshotStore
{
    public const string AppGroup = "group.com.goriparthi.redline";
    public const string FileName = "snapshot.json";
    public const string WidgetBundleID = "com.goriparthi.redline.widget";

    /// The process's own app-data copy: %APPDATA%\redline, or inside an overridden home so a
    /// test profile never finds the real machine's last reading.
    public static string? LocalAppSupportUrl => RedlineHome.IsOverridden
        ? RedlineHome.PathFor($"AppData/Roaming/redline/{FileName}")
        : Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "redline", FileName);

    /// The macOS widget container, kept so the path contract is shared. Written only if it exists.
    public static string WidgetContainerUrl =>
        RedlineHome.PathFor($"Library/Containers/{WidgetBundleID}/Data/Library/Application Support/redline/{FileName}");

    // Always available, and readable by any process of this user
    public static string UserUrl => RedlineHome.PathFor($".local/share/redline/{FileName}");

    /// App Groups do not exist on Windows, so this never resolves.
    public static string? GroupUrl(string? appGroup = AppGroup) => null;

    public static string Url(string? appGroup = AppGroup) => GroupUrl(appGroup) ?? UserUrl;

    // The widget container is included only when it exists, so nothing is created for no widget
    public static List<string> WriteTargets
    {
        get
        {
            var output = new List<string> { UserUrl };
            var container = WidgetContainerUrl;
            var parent = Path.GetDirectoryName(Path.GetDirectoryName(container));
            if (parent is not null && Directory.Exists(parent)) output.Add(container);
            if (!RedlineHome.IsOverridden && GroupUrl() is { } g) output.Add(g);
            return output;
        }
    }

    // Own container first: in a sandboxed widget that is the one guaranteed readable location
    public static List<string> ReadCandidates
    {
        get
        {
            var output = new List<string>();
            if (LocalAppSupportUrl is { } local) output.Add(local);
            if (!RedlineHome.IsOverridden && GroupUrl() is { } g) output.Add(g);
            output.Add(UserUrl);
            return output;
        }
    }

    /// Succeeds if any target takes it; a missing container must not fail the write.
    public static bool WriteEverywhere(Snapshot snapshot) =>
        WriteTargets.Select(t => Write(snapshot, t)).ToList().Contains(true);

    public static Snapshot? ReadAny()
    {
        foreach (var path in ReadCandidates)
            if (Read(path) is { } s) return s;
        return null;
    }

    public static bool Write(Snapshot snapshot, string? to = null)
    {
        var path = to ?? Url();
        string text;
        try { text = snapshot.ToJson().ToJsonString(); }
        catch
        {
            Diag.Log.Error("snapshot.encode_failed", "could not encode snapshot");
            return false;
        }
        try
        {
            Json.WriteAtomic(path, text);
            return true;
        }
        catch (Exception e)
        {
            // The widget and the CLI both read this file, so a failed write explains every stale one
            Diag.Log.Error("snapshot.write_failed", "could not write snapshot",
                new() { ["path"] = path, ["error"] = e.Message });
            return false;
        }
    }

    public static Snapshot? Read(string? from = null)
    {
        var path = from ?? Url();
        string text;
        try { text = File.ReadAllText(path); } catch { return null; }
        try
        {
            if (Json.ParseObject(text) is not { } json) throw new FormatException("not a JSON object");
            return Snapshot.FromJson(json);
        }
        catch (Exception e)
        {
            Diag.Log.Error("snapshot.decode_failed", "snapshot on disk did not decode",
                new() { ["path"] = path, ["error"] = e.Message });
            return null;
        }
    }
}
